### PR TITLE
Improved Pagination

### PR DESCRIPTION
**Pull request by GrahamCampbell**
This is just a proof of concept. I've not tested it actually works. Any feedback would be good. :)

### CODE REVIEW

File: lib/Github/ResultIterator.php [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+ * overhead due to pagination under the hood.
+ *
+ * @author Graham Campbell <graham@alt-three.com>
+ */
+class ResultIterator implements Iterator
+{
+    /**
+     * @var \Github\ResultPagerInterface
+     */
+    protected $pager;
```
Comment (stof) [inline lib/Github/ResultIterator.php]: you should make properties private IMO. It make the handling of BC easier

Comment (cursedcoder) [inline lib/Github/ResultIterator.php]: +1

**Comment by gharlan** [comment]
concept looks good :+1:

File: lib/Github/ResultIterator.php [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    /**
+     * @return array
+     */
+    protected function getNext()
+    {
+        if (!$this->pager->hasNext()) {
+            return array();
+        }
+
+        $next = $this->fetchNext();
```
Comment (gharlan) [inline lib/Github/ResultIterator.php]: missing `->pager`

File: lib/Github/ResultIterator.php [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    /**
+     * @return void
+     */
+    public function rewind()
+    {
+        $this->pos = 0;
+        $this->offset = 0;
+        $this->results = null;
```
Comment (gharlan) [inline lib/Github/ResultIterator.php]: if offset == 0 we actually do not have to reset the results.

Comment (GrahamCampbell) [inline lib/Github/ResultIterator.php]: we do need to reset them otherwise the first fetch never gets run

Comment (gharlan) [inline lib/Github/ResultIterator.php]: Maybe I'm wrong but shouldn't this work?

```php
if (0 !== $this->offset) {
    $this->results = null;
}
$this->pos = 0;
$this->offset = 0;
```

So if only one page exists, the results are only fetched once.

Comment (gharlan) [inline lib/Github/ResultIterator.php]:
```php
$it = $page->getIterator(...);

foreach ($it as $v) {
    if (some_condition()) {
        break;
    }
}

// do not fetch again if only one page exists or if the cursor still points to an entry on first page
foreach ($it as $v) {
}

Comment (GrahamCampbell) [inline lib/Github/ResultIterator.php]: I'm open to PRs to my fork. :)

File: lib/Github/ResultIterator.php [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+ *
+ * This class allows you to iterate over a large data set with a low memory
+ * overhead due to pagination under the hood.
+ *
+ * @author Graham Campbell <graham@alt-three.com>
+ */
+class ResultIterator implements Iterator
+{
+    /**
+     * @var \Github\ResultPagerInterface
```
Comment (Soullivaneuh) [inline lib/Github/ResultIterator.php]: FQN classes should be imported instead IMO.

Comment (stof) [inline lib/Github/ResultIterator.php]: no need to import anything actually: it is in the same namespace

Comment (GrahamCampbell) [inline lib/Github/ResultIterator.php]: sure - i'm just in habit of using FQCNs in phpdoc for readability - i'll change it when i have a moment

**Comment by cursedcoder** [comment]
Good concept but I can't merge it as is. @GrahamCampbell

**Pull request closed without merging**